mailpoet/mailpoet
# Introducing Action Scheduler Cron Runner [MAILPOET-4274]

This PR introduces a new method for MailPoet's background job processing that uses the Action Scheduler in the background.

This is intended as a first step in migration towards using the Action scheduler. You can read more details about the proposed process in the related P2 post (pcNwfB-NX-p2).

The idea is that the daemon runner is wrapped in a single recurring action scheduler's action. 

### How it works
There is a trigger action (`mailpoet/cron/daemon-trigger`) that in every Action Scheduler run checks whether there are jobs to process. It uses the very same method as the current trigger, which does the same in every request.

In case the trigger action detects there is a job to do it schedules a daemon run action (`mailpoet/cron/daemon-run`) and spawns a new background action scheduler runner. The deamon run action runs for 10 seconds and after the run is completed it either schedules itself again or if there is no more work to do unschedules itself. rescheduling is done automatically by the Action Scheduler

The whole logic is implemented in `MailPoet\Cron\DaemonActionSchedulerRunner`.

I tried to comment on every commit. I recommend reviewing the change commit by commit.

### Test fixes
This PR contains a bunch of fixes in tests. When working on this task I needed to enable WP-Cron for the Acceptance Tests. This caused a bunch of flaky tests. Hopefully, I fixed all of them.

### Changes in UI

The method will be selected as the default option for new installs in MailPoet > Settings > Advanced
![Screenshot 2022-05-25 at 14 51 14](https://user-images.githubusercontent.com/1082140/170266667-f98e4f7f-1d2c-437e-bea3-e805aa28dc0c.png)

There is additional status info on the MailPoet > Help > System Tab
![Screenshot 2022-07-21 at 13 20 57](https://user-images.githubusercontent.com/1082140/180201895-1cb3474e-0418-4a5d-a575-d046211405b4.png)

## Testing instructions
It would be great to test the overall functionality of the plugin with a focus on background jobs. Ideally, test on more hosts.

You need to go to Settings > Advanced and switch the task scheduler `WordPress built-in cron (recommended)`

- Does sending to a large number (over 10K) of subscribers work?
- You can test inactive subscribers' detection.
- Try scheduling some scheduled tasks to the past and observe they were executed.

### Testing build

I prepared a special build that, instead of real sending (through MSS, web host, and SMTP), only sleeps for some time and pretend the email was sent. The build can be used for testing how sending a large number of emails works without actually sending emails out.

Download >>> [mailpoet_mocked_sending.zip](https://github.com/mailpoet/mailpoet/files/9242315/mailpoet_mocked_sending.zip)

To be sure you have this build, you can check the plugin name in Plugins. It should say: `MailPoet 3 (New - Mocked Sending)`

[MAILPOET-4274]

[MAILPOET-4274]: https://mailpoet.atlassian.net/browse/MAILPOET-4274?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
